File: docs/java_patterns/SessionState.java
package com.tms.patterns.state;

/**
 * DESIGN PATTERN IMPLEMENTATION: State Pattern (Java)
 * 
 * Pattern: State (Behavioral)
 * Purpose: Alters session behavior based on internal state transitions (SCHEDULED -> COMPLETED / CANCELLED).
 */

interface SessionStateInterface {
    void complete(SessionContext context);
    void cancel(SessionContext context);
}

class ScheduledStateImpl implements SessionStateInterface {
    @Override
    public void complete(SessionContext context) {
        System.out.println("Transitioning session to COMPLETED state.");
        context.setState(new CompletedStateImpl());
    }

    @Override
    public void cancel(SessionContext context) {
        System.out.println("Transitioning session to CANCELLED state.");
        context.setState(new CancelledStateImpl());
    }
}

class CompletedStateImpl implements SessionStateInterface {
    @Override
    public void complete(SessionContext context) {
        System.out.println("Session is already completed.");
    }

    @Override
    public void cancel(SessionContext context) {
        throw new IllegalStateException("Completed session cannot be cancelled.");
    }
}

class CancelledStateImpl implements SessionStateInterface {
    @Override
    public void complete(SessionContext context) {
        throw new IllegalStateException("Cancelled session cannot be completed directly.");
    }

    @Override
    public void cancel(SessionContext context) {
        System.out.println("Session is already cancelled.");
    }
}

class SessionContext {
    private SessionStateInterface state = new ScheduledStateImpl();

    public void setState(SessionStateInterface state) {
        this.state = state;
    }

    public void complete() {
        state.complete(this);
    }

    public void cancel() {
        state.cancel(this);
    }
}

public class SessionState {
    public static void main(String[] args) {
        SessionContext session = new SessionContext();
        session.complete(); // Transitions SCHEDULED -> COMPLETED
    }
}

File: patterns/state/SessionState.js
/**
 * DESIGN PATTERN IMPLEMENTATION: State Pattern
 * 
 * Pattern: State (Behavioral)
 * Purpose: Allows a Session object to alter its behavior when its internal status 
 *          changes (SCHEDULED -> COMPLETED / CANCELLED / RESCHEDULED), enforcing valid state transitions.
 */

class ISessionState {
    complete(context) {
        throw new Error(`Cannot complete session in ${this.constructor.name} state.`);
    }

    cancel(context) {
        throw new Error(`Cannot cancel session in ${this.constructor.name} state.`);
    }

    reschedule(context) {
        throw new Error(`Cannot reschedule session in ${this.constructor.name} state.`);
    }
}

class ScheduledState extends ISessionState {
    complete(context) {
        console.log(`🔄 [State Pattern] Transitioning session ${context.id} from SCHEDULED -> COMPLETED`);
        context.setState(new CompletedState());
        return { status: 'COMPLETED', message: 'Session completed successfully.' };
    }

    cancel(context) {
        console.log(`🔄 [State Pattern] Transitioning session ${context.id} from SCHEDULED -> CANCELLED`);
        context.setState(new CancelledState());
        return { status: 'CANCELLED', message: 'Session cancelled.' };
    }

    reschedule(context) {
        console.log(`🔄 [State Pattern] Transitioning session ${context.id} from SCHEDULED -> RESCHEDULED`);
        context.setState(new RescheduledState());
        return { status: 'RESCHEDULED', message: 'Session rescheduled.' };
    }
}

class RescheduledState extends ISessionState {
    complete(context) {
        console.log(`🔄 [State Pattern] Transitioning session ${context.id} from RESCHEDULED -> COMPLETED`);
        context.setState(new CompletedState());
        return { status: 'COMPLETED', message: 'Rescheduled session completed.' };
    }

    cancel(context) {
        console.log(`🔄 [State Pattern] Transitioning session ${context.id} from RESCHEDULED -> CANCELLED`);
        context.setState(new CancelledState());
        return { status: 'CANCELLED', message: 'Rescheduled session cancelled.' };
    }

    reschedule(context) {
        return { status: 'RESCHEDULED', message: 'Session updated with new time slot.' };
    }
}

class CompletedState extends ISessionState {
    complete(context) {
        return { status: 'COMPLETED', message: 'Session is already completed.' };
    }

    cancel(context) {
        throw new Error('Completed sessions cannot be cancelled.');
    }

    reschedule(context) {
        throw new Error('Completed sessions cannot be rescheduled.');
    }
}

class CancelledState extends ISessionState {
    complete(context) {
        throw new Error('Cancelled sessions cannot be completed directly. Reschedule first.');
    }

    cancel(context) {
        return { status: 'CANCELLED', message: 'Session is already cancelled.' };
    }

    reschedule(context) {
        console.log(`🔄 [State Pattern] Transitioning session ${context.id} from CANCELLED -> RESCHEDULED`);
        context.setState(new RescheduledState());
        return { status: 'RESCHEDULED', message: 'Cancelled session reactivated and rescheduled.' };
    }
}

// Session State Context Wrapper
class SessionStateContext {
    constructor(id, currentStatusStr = 'SCHEDULED') {
        this.id = id;
        this.setStateFromStatus(currentStatusStr);
    }

    setState(stateInstance) {
        this.currentState = stateInstance;
    }

    setStateFromStatus(statusStr) {
        switch (statusStr) {
            case 'COMPLETED':
                this.currentState = new CompletedState();
                break;
            case 'CANCELLED':
                this.currentState = new CancelledState();
                break;
            case 'RESCHEDULED':
                this.currentState = new RescheduledState();
                break;
            case 'SCHEDULED':
            default:
                this.currentState = new ScheduledState();
                break;
        }
    }

    completeSession() {
        return this.currentState.complete(this);
    }

    cancelSession() {
        return this.currentState.cancel(this);
    }

    rescheduleSession() {
        return this.currentState.reschedule(this);
    }
}

module.exports = {
    SessionStateContext,
    ScheduledState,
    CompletedState,
    CancelledState,
    RescheduledState
};
